cms/core/formatting_utils.py

```diff
@@ -0,0 +1,115 @@
+    description: "StrOrPromise"
+
+
+# Type alias for cleaner function signatures
+PageDataCollection = Iterable[dict[str, Any]]
+
+
+def format_as_document_list_item(
+    title: str, url: str, content_type: "StrOrPromise", description: str
+) -> DocumentListItem:
+    """Formats an object as a list element to be used in the ONS DocumentList design system component."""
+    return {
+        "title": {"text": title, "url": url},
+        "metadata": {"object": {"text": content_type}},
+        "description": f"<p>{description}</p>",
+    }
+
+
+def _format_external_link(page_dict: dict[str, Any]) -> DocumentListItem:
+    """Format external link dictionary into DocumentListItem."""
+    return format_as_document_list_item(
+        title=page_dict["title"],
+        url=page_dict["url"],
+        content_type=_("Article"),
+        description=page_dict.get("description", ""),
+    )
+
+
+def _format_page_object(
+    page: "Page", request: Optional["HttpRequest"] = None, custom_title: Optional[str] = None
+) -> DocumentListItem:
+    """Format page object into DocumentListItem."""
+    page_datum: DocumentListItem = format_as_document_list_item(
+        title=custom_title or getattr(page, "display_title", page.title),
+        url=page.get_url(request=request),
+        content_type=getattr(page, "label", _("Page")),
+        description=getattr(page, "listing_summary", "") or getattr(page, "summary", ""),
+    )
+
+    if release_date := getattr(page, "release_date", None):
+        page_datum["metadata"]["date"] = get_document_metadata_date(release_date)
+    return page_datum
+
+
+def get_formatted_pages_list(
+    pages: PageDataCollection,
+    request: Optional["HttpRequest"] = None,
+) -> list[DocumentListItem]:
+    """Returns a formatted list of page data for the documentList DS macro.
+
+    See the search results section in https://service-manual.ons.gov.uk/design-system/components/document-list.
+    """
+    data = []
+    for page in pages:
+        if page.get("is_external", False):
+            datum = _format_external_link(page)
+        elif "internal_page" in page:
+            # Handle dict format with internal_page and optional title
+            internal_page = page["internal_page"]
+            custom_title = page.get("title")
+            datum = _format_page_object(internal_page, request, custom_title)
+        else:
+            # This should not happen in production but is a safeguard for unexpected data types
+            continue
+        data.append(datum)
+    return data
+
+
+def get_document_metadata_date(value: date | datetime, *, prefix: "StrOrPromise | None" = None) -> dict[str, Any]:
+    """Returns a dictionary with formatted date information for the DS document component metadata."""
+    return {
+        "prefix": prefix or _("Released"),
+        "showPrefix": True,
+        "iso": date_format(value, "c"),
+        "short": ons_date_format(value, "DATE_FORMAT"),
+    }
+
+
+def get_document_metadata(
```

Comment:
I think we are missing a unit for this function

Reply:
[9ba70d6](https://github.com/ONSdigital/dis-wagtail/pull/346/commits/9ba70d69f02228b53026353684b530ff357e0bb0)